### PR TITLE
Handle start and end time arithmetic overflow cases on topic subscribe

### LINKED ISSUE #521: HCS GRPC Server throwing ArithmeticException: long overflow errors
Over the weekend during normal and non-peak operations of the tesnet the GRPC module triggered an alert that was caused by an ArithmeticException: long overflow.

This occurred about 3 times
https://hedera.pagerduty.com/incidents/P0VGW2Z

The error seems to come from the InstantToLongConverter.convert() which takes an Instant and converts it to a Long.
This happens when the startTime and endTime of a TopicMessageFilter object are converted to longs when creation the DB whereClause Criteria object in TopicMessageRepositoryCustomImpl.findByFilter()

The code base has tests that confirm this logic and as such it's by design.
Will confirm that no action needs to be taken and there's no unintended impact.

At the least if this is expected we shouldn't alert on it.

-----
error encountered

```
Feb 01 19:32:13 mirrornode-testnet-grpc-1 java[816]: 2020-02-01 19:32:13,952 INFO  [grpc-default-executor-725] c.h.m.g.s.TopicMessageServiceImpl Topic 0.159463 listen complete with 9 messages in 23.18 s
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at java.lang.Thread.run(Thread.java:834) [?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) [grpc-core-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [grpc-core-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:814) [grpc-core-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331) [grpc-core-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.Contexts$ContextualizedServerCallListener.onHalfClose(Contexts.java:86) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35) [grpc-api-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:172) [grpc-stub-1.25.0.jar!/:1.25.0]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.hedera.mirror.api.proto.ReactorConsensusServiceGrpc$MethodHandlers.invoke(ReactorConsensusServiceGrpc.java:116) [hedera-mirror-protobuf-0.5.1.jar!/:0.5.1]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.salesforce.reactorgrpc.stub.ServerCalls.oneToMany(ServerCalls.java:67) [reactor-grpc-stub-1.0.0.jar!/:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.Flux.subscribeWith(Flux.java:8307) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.Flux.subscribe(Flux.java:8143) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:54) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onSubscribe(MonoFlatMapMany.java:134) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2186) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onNext(MonoFlatMapMany.java:156) [reactor-core-3.3.1.RELEASE.jar!/:3.3.1.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.hedera.mirror.grpc.service.TopicMessageServiceImpl$$EnhancerBySpringCGLIB$$18960421.subscribeTopic(<generated>) ~[classes!/:0.5.1]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:120) ~[spring-context-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.hedera.mirror.grpc.service.TopicMessageServiceImpl$$FastClassBySpringCGLIB$$40d0042f.invoke(<generated>) ~[classes!/:0.5.1]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.hedera.mirror.grpc.service.TopicMessageServiceImpl.subscribeTopic(TopicMessageServiceImpl.java:52) ~[classes!/:0.5.1]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.sun.proxy.$Proxy97.findByFilter(Unknown Source) ~[?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:605) ~[spring-data-commons-2.2.3
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:621) ~[spring-data-commons-2.2
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.2.2.RELEASE.jar!/:5.2.2.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:657) ~[spring-data-com
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:204) ~[spring-data-commons-2.2.3.RELEASE.jar!/:2.2.3.RELEASE]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:371) ~[spring-data-commons-2.2.3.RELEASE.jar!/:2.
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at jdk.internal.reflect.GeneratedMethodAccessor39.invoke(Unknown Source) ~[?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.hedera.mirror.grpc.repository.TopicMessageRepositoryCustomImpl.findByFilter(TopicMessageRepositoryCustomImpl.java:48) ~[classes!/:0.5.1]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at com.hedera.mirror.grpc.converter.InstantToLongConverter.convert(InstantToLongConverter.java:40) ~[classes!/:0.5.1]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]:         at java.lang.Math.multiplyExact(Math.java:946) ~[?:?]
Feb 01 19:32:03 mirrornode-testnet-grpc-1 java[816]: java.lang.ArithmeticException: long overflow
```

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:

When a client subscribes to a topic they provide a startTime and an optional endTime instant. These are converted to long to be used in the db query to compare against the consensus time stamp to determine which messages are filtered in or out.

Currently instants out of range i.e. less than EPOCH or results in long values created than max will result in an Arithmetic long overflow exception.

To resolve this issue we need to account for these edge ranges. As such I've

- Added a convertOrDefault() method to the InstantToLongConverter that will scope instants in to the manageable range
- Updated TopicMessageRepositoryCustomImpl to call this for startTime
- Updated TopicMessageRepositoryCustomImpl to check for a > than 0 long value to specify endTime
- Updated appropriate tests to verify scenario

**Which issue(s) this PR fixes**:
Fixes #521 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java [hunk 1]
```diff
@@ -121,6 +120,22 @@ void findByFilterWithStartTime() {
                 .verifyComplete();
     }
 
+    @Test
+    void findByFilterWithStartTimeOverflow() {
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java]: Why have one test for just start time overflow and another test that has both start and end time overflow? I would say the latter is sufficient.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java [hunk 2]
```diff
@@ -47,11 +47,12 @@
                 .and("topic_num")
                 .is(filter.getTopicNum())
                 .and("consensus_timestamp")
-                .greaterThanOrEquals(instantToLongConverter.convert(filter.getStartTime()));
+                .greaterThanOrEquals(instantToLongConverter.convertOrDefault(filter.getStartTime()));
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: My concern here is this approach has to be done for every query instead of just once inside `ConsensusController` to validate input. Exception generation and catching are orders of magnitude slower then normal java functions. And queries are now split into pages and run every 2s times Y subscribers, so might have more impact.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java [hunk 3]
```diff
@@ -47,11 +47,12 @@
...
 
-        if (filter.getEndTime() != null) {
+        Long endTime = instantToLongConverter.convertOrDefault(filter.getEndTime());
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: This approach does not address all instances of InstantToLongConverter since the shared poller uses it as well. Would be better to centralize in controller.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/converter/InstantToLongConverter.java [hunk 4]
```diff
@@ -39,4 +39,12 @@ public Long convert(Instant instant) {
         }
         return Math.addExact(Math.multiplyExact(instant.getEpochSecond(), NANOS_PER_SECOND), instant.getNano());
     }
+
+    public Long convertOrDefault(Instant instant) {
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/converter/InstantToLongConverter.java]: Not a fan of the extra method. The class implements Spring's Converter interface so it can be registered with Spring and used automatically within the framework. This extra method bypasses that. Though to be fair we are actually just calling it manually since Spring R2DBC doesn't currently support this, but in the future it could and we wouldn't need to call this manually in the repository.

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java [hunk 5]
```diff
@@ -106,7 +106,6 @@ void findByFilterWithTopicNum() {
 
     @Test
     void findByFilterWithStartTime() {
-        TopicMessage topicMessage1 = domainBuilder.topicMessage().block();
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java]: The line was unused. I didn't change the test logic

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java]: Sure.

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryTest.java]: It's not unused. That line inserts into the database, validating that rows before start time are not returned. Please put it back.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: Makes sense to centralize it's call in the controller.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: Agreed. Will do

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 2 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: Exception logic is fine if just done once, just not in a loop. Other approach would be to create an Instant that represents our max supported and just compare with isAfter/isBefore.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/converter/InstantToLongConverter.java (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/converter/InstantToLongConverter.java]: Removing extra method. Agreed

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: Will handle centralization in next commit as it would include modifying TopicMessageFilter class

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: This was what I ended up doing

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: I see the centralization benefit about converting once. 
To centralize however, it would require both TopicMessageFilter and TopicMessage times be converted to longs.
The result is all future arithmetic on them will be in the form of longs that represent nanoseconds and not utilizing the benefit on Instant object operations.
I think leaving it as Instant leaves it easier to read and operate on the time in a way that's clear.
Let me know if you feel strongly about centralizing or if I missed your intent here.

Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: Why would it require both TopicMessageFilter and TopicMessage times be converted to longs? Just check the proto timestamp if the seconds or nanos is less than 0 then set `messageFilter.setStartTime(Instant.EPOCH)` if so.  Likewise check if seconds+nanos > max that can fit in our long and skip setting filter.endTime if so.

Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: I think I see the mixup.
Your centralization comment suggested you wanted there to be a central location for the convert() call. So that all the listeners would not have to call it as they do now.
What you've suggested above was a consideration but doesn't achieve centralization of the convert() call.
Doing the above will still require the InstantToLong conversion prior to the dbclient operation.
My last comment explained the effect of centralizing the convert() call in the controller and therefore having to handle times represented as long all the way to the repository call.

I think what you meant was centralizing the overflow catch logic, which makes sense.
I think we're on the same page now.
I'll make the change

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 6]
```diff
@@ -49,6 +51,8 @@
 @RequiredArgsConstructor
 public class ConsensusController extends ReactorConsensusServiceGrpc.ConsensusServiceImplBase {
 
+    private static final long LONG_MAX_SECONDS = 9_223_372_036_000_000_000L;
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: nit: `Long.MAX_VALUE / NANOS_PER_SECOND * NANOS_PER_SECOND` might be clearer

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 7]
```diff
@@ -72,11 +76,23 @@ private TopicMessageFilter toFilter(ConsensusTopicQuery query) {
...
+                builder.startTime(ProtoUtil.fromTimestamp(startTimeStamp));
+            }
         }
 
         if (query.hasConsensusEndTime()) {
-            builder.endTime(ProtoUtil.fromTimestamp(query.getConsensusEndTime()));
+            Timestamp endTimeStamp = query.getConsensusEndTime();
+
+            // only set endTime if it's smaller than the seconds and nanoseconds in the Long.MAX value
+            if (endTimeStamp.getSeconds() <= LONG_MAX_SECONDS && endTimeStamp.getNanos() <= LONG_MAX_NANOSECONDS) {
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: This is wrong.  Having  seconds=0 and nanos=999999999 will cause end time to not be set.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 8]
```diff
@@ -72,11 +76,23 @@ private TopicMessageFilter toFilter(ConsensusTopicQuery query) {
                 .topicNum((int) query.getTopicID().getTopicNum());
 
         if (query.hasConsensusStartTime()) {
-            builder.startTime(ProtoUtil.fromTimestamp(query.getConsensusStartTime()));
+            Timestamp startTimeStamp = query.getConsensusStartTime();
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Please add a test to `ConsensusControllerTest` for these scenarios.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 9]
```diff
@@ -49,6 +51,8 @@
 @RequiredArgsConstructor
 public class ConsensusController extends ReactorConsensusServiceGrpc.ConsensusServiceImplBase {
 
+    private static final long LONG_MAX_SECONDS = 9_223_372_036_000_000_000L;
+    private static final int LONG_MAX_NANOSECONDS = 854_775_807;
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: nit: `Long.MAX_VALUE % NANOS_PER_SECOND` might be clearer

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Good point. Changed

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java (hunk 9 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Good point. Changed

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java (hunk 8 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Added two, pre epoch and post max in the COntrollerTest

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Good catch. I'm just gonna do a boundary check to ensure all cases

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 10]
```diff
@@ -104,4 +121,13 @@ private Throwable unknownError(Throwable t) {
         log.error("Unknown error subscribing to topic", t);
         return Status.INTERNAL.augmentDescription(t.getMessage()).asRuntimeException();
     }
+
+    private boolean validEndTimeStamp(Timestamp endTimeStamp) {
+        if (endTimeStamp == null) {
+            return false;
+        }
+
+        return (endTimeStamp.getSeconds() >= 0 && endTimeStamp.getSeconds() <= LONG_MAX_SECONDS) &&
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: This should fail if endTimestamp is negative instead of assuming they want to never end

Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Addressed

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 11]
```diff
@@ -72,11 +74,23 @@ private TopicMessageFilter toFilter(ConsensusTopicQuery query) {
                 .topicNum((int) query.getTopicID().getTopicNum());
 
         if (query.hasConsensusStartTime()) {
-            builder.startTime(ProtoUtil.fromTimestamp(query.getConsensusStartTime()));
+            Timestamp startTimeStamp = query.getConsensusStartTime();
+
+            // scope pre epoch timestamps to epoch instant
+            if (TimestampUtil.isValidTimeStamp(startTimeStamp)) {
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: This looks like it will throw an IAE if startTimeStamp is negative, but we said negative timestamps should just default to epoch.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        if (endTimeStamp.getSeconds() < 0 || endTimeStamp.getNanos() < 0) {
+            log.warn("Negative endTimeStamp supplied");
+            throw Status.INVALID_ARGUMENT.augmentDescription("Negative endTimeStamp supplied").asRuntimeException();
+        }
+
+        // valid if seconds is less than max or if seconds is at max and nanoseconds are at or below max
+        // 0 <= valid_time < LONG_MAX_SECONDS (9_223_372_036_000_000_000L) + LONG_MAX_NANOSECONDS (854_775_807)
+        return (endTimeStamp.getSeconds() < LONG_MAX_SECONDS) ||
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java]: Not sure naming makes sense. A timestamp can be valid if greater than LONG_MAX_SECONDS, it's just not valid when we convert it to our internal representation of it as a long. It may make more sense to just move this logic to `ProtoUtils.fromTimestamp()`. A parameter can be added to control whether to take timestamp as is or do bounds checking.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+@Log4j2
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class TimestampUtil {
+    private static final long NANOS_PER_SECOND = 1_000_000_000L;
+    public static final long LONG_MAX_SECONDS = Long.MAX_VALUE / NANOS_PER_SECOND * NANOS_PER_SECOND;
+    public static final int LONG_MAX_NANOSECONDS = (int) (Long.MAX_VALUE % NANOS_PER_SECOND);
+
+    public static boolean isValidTimeStamp(Timestamp endTimeStamp) {
+        if (endTimeStamp == null) {
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java]: Shouldn't be named endTimeStamp

Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java]: File will be removed.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java (hunk 11 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: That's a regression. startTimes out of the bound should indeed be bounded to EPOCH as we agreed and no exception thrown. That case should be for endTime only.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/util/TimestampUtil.java]: Moved this logic to ProtoUtil.
I didn't like the extra parameter option as in the startTime case you want to check and then covert or scope . Then in the endTime case you want to check and throw and error for the client if it's out of bound.
As such it makes sense to have a check method (say isLongSupportedTimeStamp() ) and based in its result take the appropriate action for your given timestamp.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java [hunk 14]
```diff
@@ -72,11 +74,25 @@ private TopicMessageFilter toFilter(ConsensusTopicQuery query) {
...
 
         if (query.hasConsensusStartTime()) {
-            builder.startTime(ProtoUtil.fromTimestamp(query.getConsensusStartTime()));
+            Timestamp startTimeStamp = query.getConsensusStartTime();
+            Instant startInstant = ProtoUtil.fromTimestamp(startTimeStamp);
+
+            if (startInstant.isBefore(Instant.EPOCH)) {
+                builder.startTime(Instant.EPOCH);
+            } else {
+                builder.startTime(ProtoUtil.fromTimestamp(startTimeStamp));
```
Comment (steven-sheehy) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/controller/ConsensusController.java]: Don't need to re-parse timestamp. Also can reduce if else to ternary: 
`builder.startTime(startInstant.isBefore(Instant.EPOCH) ? Instant.EPOCH : startInstant);`
